pkg/grpc/requestinfo/requestinfo_test.go

```diff
@@ -0,0 +1,206 @@
+	pb "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/pkg/contextutil"
+	"github.com/stackrox/rox/pkg/netutil/pipeconn"
+	"github.com/stackrox/rox/pkg/testutils"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/peer"
+)
+
+var insecureTLSConfig = &tls.Config{InsecureSkipVerify: true}
+var insecureSkipVerify = credentials.NewTLS(insecureTLSConfig)
+
+const userAgentKey = "User-Agent"
+
+func makeTestRequest(url string) *http.Request {
+	req, _ := http.NewRequest("GET", url, nil)
+	req.Header.Add("test-key", "test value")
+	req.Header.Add(userAgentKey, "test agent")
+	return req
+}
+
+func testRequest(t *testing.T, req *HTTPRequest) {
+	require.NotNil(t, req)
+	assert.Equal(t, "GET", req.Method)
+	assert.Equal(t, "test value", req.Headers.Get("test-key"))
+	assert.Equal(t, "test agent", req.Headers.Get(userAgentKey))
+}
+
+type pingService struct {
+	pb.UnimplementedPingServiceServer
+	ri *RequestInfo
+}
+
+func (s *pingService) Ping(ctx context.Context, in *pb.Empty) (*pb.PongMessage, error) {
+	ri := FromContext(ctx)
+	s.ri = &ri
+	return &pb.PongMessage{Status: "pong"}, nil
+}
+
+func Test_PureGRPC(t *testing.T) {
+	handler := NewRequestInfoHandler()
+
+	grpcServer := grpc.NewServer(grpc.UnaryInterceptor(
+		contextutil.UnaryServerInterceptor(handler.UpdateContextForGRPC)))
+
+	serviceInstance := &pingService{}
+	pb.RegisterPingServiceServer(grpcServer, serviceInstance)
+
+	server := httptest.NewUnstartedServer(handler.HTTPIntercept(http.HandlerFunc(grpcServer.ServeHTTP)))
+	server.EnableHTTP2 = true
+	server.StartTLS()
+	defer server.Close()
+
+	conn, err := grpc.Dial(server.Listener.Addr().String(),
+		grpc.WithTransportCredentials(insecureSkipVerify))
+	require.NoError(t, err)
+	defer func() { _ = conn.Close() }()
+
+	c := pb.NewPingServiceClient(conn)
+	resp, err := c.Ping(context.Background(), &pb.Empty{})
+	require.NoError(t, err)
+	require.Equal(t, "pong", resp.Status)
+
+	ri := serviceInstance.ri
+	require.NotNil(t, ri)
+	require.Nil(t, ri.HTTPRequest)
+	assert.NotNil(t, ri.Metadata)
+	assert.Equal(t, []string{"application/grpc"}, ri.Metadata.Get("content-type"))
+	assert.Contains(t, ri.Metadata.Get(userAgentKey)[0], "grpc-go/")
+}
+
+func Test_gRPCGateway(t *testing.T) {
+	handler := NewRequestInfoHandler()
+	serviceInstance := &pingService{}
+
+	cert := testutils.IssueSelfSignedCert(t, "*.example.com", "*.example.com")
+	creds := credentials.NewTLS(&tls.Config{
+		InsecureSkipVerify: true,
+		Certificates:       []tls.Certificate{cert},
+	})
+
+	grpcServer := grpc.NewServer(grpc.Creds(creds),
+		grpc.UnaryInterceptor(contextutil.UnaryServerInterceptor(
+			handler.UpdateContextForGRPC)))
+	pb.RegisterPingServiceServer(grpcServer, serviceInstance)
+
+	lis, dialContext := pipeconn.NewPipeListener()
+	go func() {
+		if err := grpcServer.Serve(lis); err != nil {
+			log.Fatal(err)
+		}
+	}()
```

Comment:
This could lead to flakes, is there some way to ensure we only start once the server is already running / serving?

Reply:
As the connection is established via non-buffered pipe, the client will wait for the server to listen.
I tested it by adding a timeout before calling `Serve()`, and the test passes.